crates/ty_python_semantic/src/types/generics.rs

```diff
@@ -723,6 +723,23 @@ impl<'db> SpecializationBuilder<'db> {
                 }
             }
 
+            (Type::Callable(formal), Type::FunctionLiteral(actual)) => {
+                let formal_signature = formal.signatures(self.db);
+                let actual_signature = actual.signature(self.db);
+
+                let formal_return_types: Vec<_> = formal_signature.overloads.iter().filter_map(|signature| signature.return_ty).collect();
+                let actual_return_types: Vec<_> = actual_signature.overloads.iter().filter_map(|signature| signature.return_ty).collect();
```

Comment:
I think the comments above suggest that the overload handling here should be a bit different. We should probably try inferring each overload in order against the actual signature, and go with the first one that matches. That definitely increases the complexity of this PR, though.